staging/vhost-device-sound/src/audio_backends/pipewire.rs

```diff
@@ -331,54 +344,106 @@ impl AudioBackend for PwBackend {
+                                } else {
+                                    0
+                                };
                             }
-                            let p = &mut slice[0..n_bytes];
-                            if avail <= 0 {
-                                // SAFETY: We have assured above that the pointer is not null
-                                // safe to zero-initialize the pointer.
-                                unsafe {
-                                    // pad with silence
-                                    ptr::write_bytes(p.as_mut_ptr(), 0, n_bytes);
-                                }
-                            } else {
-                                // consume() always reads (buffer.desc_len() - buffer.pos) bytes
-                                buffer.consume(p).expect("failed to read buffer from guest");
-
-                                start += n_bytes;
-
-                                buffer.pos = start;
-
-                                if start >= buffer.desc_len() as usize {
-                                    streams.buffers.pop_front();
-                                }
+                            spa::Direction::Output => {
+                                let datas = buf.datas_mut();
+                                let frame_size = info.channels * size_of::<i16>() as u32;
+                                let data = &mut datas[0];
+                                let n_bytes = if let Some(slice) = data.data() {
+                                    let mut n_bytes = slice.len();
+                                    let mut streams = streams.write().unwrap();
+                                    let streams = streams
+                                        .get_mut(stream_id as usize)
+                                        .expect("Stream does not exist");
+                                    let Some(buffer) = streams.buffers.front_mut() else {
+                                        return;
+                                    };
+
+                                    let mut start = buffer.pos;
+
+                                    let avail = (buffer.desc_len() - start as u32) as i32;
+
+                                    if avail < n_bytes as i32 {
+                                        n_bytes = avail.try_into().unwrap();
+                                    }
+                                    let p = &mut slice[0..n_bytes];
+                                    if avail <= 0 {
+                                        // SAFETY: We have assured above that the pointer is not
+                                        // null
+                                        // safe to zero-initialize the pointer.
+                                        unsafe {
+                                            // pad with silence
+                                            ptr::write_bytes(p.as_mut_ptr(), 0, n_bytes);
+                                        }
```

Comment:
Question: why are you unsafely memset'ing a pointer to a rust slice instead of using `slice[0..n_bytes].fill(0)`? It's not wrong, just unnecessary.

Reply:
I was thinking C like, I agree this is much safer.
But this change is not for this PR.

Reply:
Indeed, I think you should fix this first separately instead of adding this new code and then fixing it.

Reply:
Nevermind, thought this was in the Capture code. You're right, not for this PR.